// @8man/react-native-media-console/lib/commonjs/components/Seekbar/Seekbar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Seekbar = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _styles = require("./styles");
var _utils = require("@8man/react-native-media-console/src/utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const Seekbar = ({
  seekColor,
  seekerFillWidth,
  seekerPosition,
  cachedPosition,
  seekerPanHandlers,
  setSeekerWidth,
  showDuration,
  showHours,
  showTimeRemaining,
  time,
  duration
}) => {
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      marginBottom: 55,
      flexDirection: 'row',
      justifyContent: 'center',
      alignItems: 'flex-end'
    }
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, null, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: {
      color: 'hsl(0, 0%, 70%)'
    }
  }, (0, _utils.formatTime)({
    duration,
    time,
    showDuration,
    showHours,
    showTimeRemaining
  }))), /*#__PURE__*/_react.default.createElement(_reactNative.View, _extends({
    style: {
      ..._styles.styles.container,
      width: '80%'
    },
    collapsable: false
  }, seekerPanHandlers), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.styles.track,
    onLayout: event => setSeekerWidth(event.nativeEvent.layout.width),
    pointerEvents: 'none'
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [{
      width: cachedPosition,
      backgroundColor: '#dedede',
      height: 4,
      position: 'absolute',
      top: 0,
      borderRadius: 3
    }],
    pointerEvents: 'none'
  }), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [_styles.styles.fill, {
      width: seekerFillWidth,
      backgroundColor: seekColor || '#FFF'
    }],
    pointerEvents: 'none'
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [_styles.styles.handle, {
      left: seekerPosition
    }],
    pointerEvents: 'none'
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [_styles.styles.circle, {
      backgroundColor: seekColor || '#FFF'
    }],
    pointerEvents: 'none'
  }))), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {}
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: {
      color: 'hsl(0, 0%, 70%)'
    }
  }, (0, _utils.formatTime)({
    duration,
    time: duration,
    showDuration,
    showHours,
    showTimeRemaining
  }))));
};
exports.Seekbar = Seekbar;
