// @8man/react-native-media-console/lib/commonjs/components/Title.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Title = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _styles2 = require("./styles");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const Title = title => {
  if (title) {
    return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: [_styles2.styles.control, _styles.title]
    }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
      style: [_styles2.styles.text, _styles.titleText],
      numberOfLines: 1
    }, title.primary), title.secondary && /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
      style: [_styles.secondaryText, {
        color: 'hsl(0, 0%, 70%)'
      }],
      numberOfLines: 1
    }, title.secondary));
  }
  return null;
};
exports.Title = Title;
const _styles = _reactNative.StyleSheet.create({
  title: {
    alignItems: 'flex-start',
    flex: 1,
    flexDirection: 'column',
    padding: 0,
    marginRight: 16,
    marginTop: 15,
    height: 37
  },
  titleText: {
    textAlign: 'center'
  },
  secondaryText: {
    fontSize: 12
  }
});
