// @8man/react-native-media-console/lib/commonjs/components/PlayPause/PlayPause.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.playPauseRef = exports.PlayPause = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _Control = require("../Control");
var _NullControl = require("../NullControl");
var _styles = require("./styles");
var _Loader = require("@8man/react-native-media-console/src/components/Loader");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); } // @ts-nocheck
const playPauseRef = exports.playPauseRef = /*#__PURE__*/(0, _react.createRef)();
const play = require('../../assets/img/playNew.png');
const pause = require('../../assets/img/pauseNew.png');
const rewind = require('../../assets/img/rewind10.png');
const forward = require('../../assets/img/forward10.png');
const PlayPause = ({
  animations: {
    AnimatedView,
    ...animations
  },
  disablePlayPause,
  disableSeekButtons,
  paused,
  // pauseLabel,
  buffering,
  togglePlayPause,
  resetControlTimeout,
  showControls,
  onPressForward,
  onPressRewind,
  primaryColor
}) => {
  const animatedStyles = {
    zIndex: showControls ? 99999 : 0
  };
  if (disablePlayPause) {
    return /*#__PURE__*/_react.default.createElement(_NullControl.NullControl, null);
  }
  return /*#__PURE__*/_react.default.createElement(AnimatedView, {
    pointerEvents: 'box-none',
    style: [_styles.styles.container, animatedStyles, animations.controlsOpacity]
  }, !disableSeekButtons ? /*#__PURE__*/_react.default.createElement(_Control.Control, {
    disabled: !showControls,
    callback: onPressRewind,
    resetControlTimeout: resetControlTimeout
  }, /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    onPress: onPressRewind
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    source: rewind,
    style: _styles.styles.rewind
  }))) : null, /*#__PURE__*/_react.default.createElement(_Control.Control, _extends({
    disabled: !showControls,
    callback: togglePlayPause,
    resetControlTimeout: resetControlTimeout,
    style: _styles.styles.playContainer,
    controlRef: playPauseRef
  }, _reactNative.Platform.isTV ? {
    hasTVPreferredFocus: showControls
  } : {}), buffering ? /*#__PURE__*/_react.default.createElement(_Loader.Loader, {
    color: primaryColor
  }) : /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    onPress: togglePlayPause
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    source: paused ? play : pause,
    style: _styles.styles.play
  }))), !disableSeekButtons ? /*#__PURE__*/_react.default.createElement(_Control.Control, {
    disabled: !showControls,
    callback: onPressForward,
    resetControlTimeout: resetControlTimeout
  }, /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    onPress: onPressForward
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    source: forward,
    style: _styles.styles.forward
  }))) : null);
};
exports.PlayPause = PlayPause;
