// @8man/react-native-media-console/lib/module/components/Control.js
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
import React, { useState } from 'react';
import { TouchableHighlight } from 'react-native';
import { styles } from './styles';
export const Control = ({
  children,
  callback,
  controlRef,
  disabled,
  style = {},
  ...props
}) => {
  const [focused, setFocused] = useState(false);
  const setFocusedState = () => setFocused(true);
  const cancelFocusedState = () => setFocused(false);
  const focusedStyle = focused ? {
    opacity: 1
  } : {};
  return /*#__PURE__*/React.createElement(TouchableHighlight, _extends({
    onFocus: setFocusedState,
    onBlur: cancelFocusedState,
    disabled: disabled,
    ref: controlRef,
    underlayColor: "transparent",
    activeOpacity: 1,
    onPress: () => {
      callback && callback();
    },
    style: [styles.control, style, focused && focusedStyle]
  }, props), children);
};
